psi4/src/psi4/libfock/snLinK.cc

```diff
@@ -0,0 +1,498 @@
+    using prim_array = typename GauXC::Shell<T>::prim_array;
+    using cart_array = typename GauXC::Shell<T>::cart_array;
+
+    GauXC::BasisSet<T> gauxc_basisset(psi4_basisset->nshell());
+ 
+    for (size_t ishell = 0; ishell != psi4_basisset->nshell(); ++ishell) {
+        auto psi4_shell = psi4_basisset->shell(ishell);
+       
+        const auto nprim = GauXC::PrimSize(psi4_shell.nprimitive());
+        prim_array alpha; 
+        prim_array coeff;
+
+        for (size_t iprim = 0; iprim != psi4_shell.nprimitive(); ++iprim) {
+            alpha.at(iprim) = psi4_shell.exp(iprim);
+            coeff.at(iprim) = psi4_shell.coef(iprim);
+        }
+
+        auto psi4_shell_center = psi4_shell.center();
+        cart_array center = { psi4_shell_center[0], psi4_shell_center[1], psi4_shell_center[2] };
+
+        gauxc_basisset[ishell] = GauXC::Shell(
+            nprim,
+            GauXC::AngularMomentum(psi4_shell.am()), 
+            (force_cartesian ? GauXC::SphericalType(false) : GauXC::SphericalType( !(psi4_shell.is_cartesian()) ) ),
+            alpha,
+            coeff,
+            center,
+            false // do not normalize shell via GauXC; it is normalized via Psi4
+        );
+    }
+    
+    for (auto& sh : gauxc_basisset) {
+        sh.set_shell_tolerance(basis_tol); 
+    }
+
+    return gauxc_basisset;
+}
+
+snLinK::snLinK(std::shared_ptr<BasisSet> primary, Options& options) : SplitJK(primary, options) {
+    timer_on("snLinK: Setup");
+    
+    // => Part #1: Options Processing <= //
+    timer_on("snLinK: Options Processing");
+
+    // set a few Psi4-specific parameters
+    incfock_iter_ = false;
+  
+    basis_tol_ = options.get_double("SNLINK_BASIS_TOLERANCE");
+
+    pruning_scheme_ = options_.get_str("SNLINK_PRUNING_SCHEME");
+    radial_scheme_ = options_.get_str("SNLINK_RADIAL_SCHEME");
+ 
+//#if psi4_SHGSHELL_ORDERING == LIBINT_SHGSHELL_ORDERING_STANDARD
+#if psi4_SHGSHELL_ORDERING == 1 
+    is_cca_ = true;
+//#elif psi4_SHGSHELL_ORDERING == LIBINT_SHGSHELL_ORDERING_GAUSSIAN
+#elif psi4_SHGSHELL_ORDERING == 2 
+    is_cca_ = false;
+#else
+    #error "unknown value of macro psi4_SHGSHELL_ORDERING"
+#endif
+
+    format_ = Eigen::IOFormat(4, 0, ", ", "\n", "[", "]");
+    
+    // create mappings for GauXC pruning and radial scheme enums
+    auto [ pruning_scheme_map, radial_scheme_map ] = generate_enum_mappings();
+
+    // define runtime environment and execution space
+    use_gpu_ = options_.get_bool("SNLINK_USE_GPU"); 
+    auto ex = use_gpu_ ? GauXC::ExecutionSpace::Device : GauXC::ExecutionSpace::Host;  
+
+    std::unique_ptr<GauXC::RuntimeEnvironment> rt = nullptr; 
+#ifdef GAUXC_HAS_DEVICE 
+    if (use_gpu_) {
+        // 0.9 indicates to use maximum 90% of maximum GPU memory, I think?
+        rt = std::make_unique<GauXC::DeviceRuntimeEnvironment>( GAUXC_MPI_CODE(MPI_COMM_WORLD,) 0.9 );
```

Comment:
Perhaps making this 90% an expert option would be a good idea.

Reply:
Yes, this is actually dangerous if there are other GPU libs in flight - the more sustainable solution would eventually be to allocate a (common) buffer and pass it in - this mode is supported to enable interop with other GPU libraries that support external allocations (to the largest extent possible) - GPU data occupancy is assumed to be ephemeral, allocation is just amortized as it's expensive in critical paths. If it's desirable, we could also set up a mode where the GPU memory is reallocated every instance of `eval_xyz` to allow for (sensible) interop with libraries that don't support external memory allocation (not sure what BrianQC does or does not support)

Reply:
Agreed on @TiborGY's suggestion; I did start implementing your idea. 

I can definitely see the concerns that @wavefunction91 brings up - I presume there could be issues with different GPU libraries allocating GPU memory simultaneously. But I have to wonder if the common buffer idea falls within the scope of this PR, even though it is likely the long-term solution. If we do think it should be added this PR, I can dig around and see if there are any other GPU-enabled Psi4 capabilities (e.g., perhaps the GPU CC plugins) and look at their use of external GPU memory allocation, or lack thereof. Thoughts?

Reply:
If there isn't pervasive GPU code in Psi4, I probably wouldn't bog this down with it. Going with @TiborGY 's suggestion may just be the way to go. IIRC the GPU CC modules allocate memory to themselves, so as long as the XC instance goes out of scope before a CC calculation starts, you're probably fine. I'm not sure if that module can use a preallocated buffer.

Reply:
So for now, I have implemented @TiborGY's suggestion. There is a new keyword for snLinK, `SNLINK_GPU_MEM`, which controls the % of GPU memory used by the GauXC instance (defaulting to 90, the previously-hardcoded behavior). I do want to explicitly test this out with the GPU CC plugin, however, which I have not yet done.

Reply:
So I dug around the GPU DFCC plugin a bit. It seems to be hardcoded to only work with a limited subset of `SCF_TYPES`  (`CD, DISK_DF`), so you can't currently run GPU DFCC plugin in conjunction with `SNLINK` anyway. I suppose because of that, there shouldn't be issues between those two in practice currently?